### PR TITLE
best effort to label namespace when starting new workspace

### PR DESCRIPTION
**Pull request by sparkoo**
Signed-off-by: Michal Vala <mvala@redhat.com>

### What does this PR do?
When starting new workspace, try label the namespace, if che SA have permissions to do so.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17504

### How to test this PR?

1. Ensure that che SA has ClusterRole with `get` and `update` `namespace` permission. You can use this:
```
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: che-extraperm
rules:
- apiGroups: [""]
  resources: ["namespaces"]
  verbs: ["update", "get"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: che-extraperm
subjects:
- kind: ServiceAccount
  name: che
  namespace: che
roleRef:
  kind: ClusterRole
  name: che-extraperm
  apiGroup: rbac.authorization.k8s.io
```
2. Create a new workspace
3. Check that namespace where workspace has started is labeled with labels from `che.infra.kubernetes.namespace.labels` property

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by skabashnyuk** [comment]
looks good to me in general.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 1]
```diff
@@ -127,7 +142,45 @@ void prepare(boolean canCreate) throws InfrastructureException {
...
+          "Nothing to do, namespace [{}] already has all required labels.",
+          namespace.getMetadata().getName());
+      return;
+    }
+
+    newLabels.putAll(ensureLabels);
+    namespace.getMetadata().setLabels(newLabels);
+
+    try {
+      cheClientFactory.create().namespaces().createOrReplace(namespace);
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: I have some doubts that we need to update labels for the namespace where che server is running. Maybe it's the job of the operator or another deployment tool.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 2]
```diff
@@ -127,7 +142,45 @@ void prepare(boolean canCreate) throws InfrastructureException {
...
+
+    if (newLabels.entrySet().containsAll(ensureLabels.entrySet())) {
+      LOG.debug(
+          "Nothing to do, namespace [{}] already has all required labels.",
+          namespace.getMetadata().getName());
+      return;
+    }
+
+    newLabels.putAll(ensureLabels);
+    namespace.getMetadata().setLabels(newLabels);
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: Do we have a good failover here? What will happen if this operation is not allowed?

Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: this is just a setter, it's not updating anything on cluster. the critical operation is `cheClientFactory.create().namespaces().createOrReplace(namespace);` where we catch the exception

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 1 above)
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: it is updating the namespace passed to this method -> `createOrReplace(namespace)`. The `cheClientFactory` is just creating the kubernetes client with `che` SA.
I should probably name the variables better or comment, so it's more clear what this is doing.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 2 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: got it

**Comment by sparkoo** [comment]
[crw-ci-test]

**Comment by dmytro-ndp** [comment]
@sparkoo: FYI: Happy path tests are failing because of new problem with build of Che server image https://github.com/eclipse/che/issues/18284

**Comment by sparkoo** [comment]
[ci-build]

**Review by skabashnyuk — APPROVED** [review]

**Review by metlos — APPROVED** [review]

**Pull request merged**